### PR TITLE
[closed] Update player info encumbrance penalties to match the actual penalties

### LINKED ISSUE #16833: The encumbrance penalties shown in Player Info don't match the real ones
Almost all of the encumbrance penalties shown under View Player Info > Encumbrance and Warmth are either out of sync with the actual penalties or are just a sentence with no numbers in it.

In the long run all the encumbrance code should probably be changed to share functions with player::disp_info() to keep this from happening again.

### PR DESCRIPTION
**Pull request by ghost**
Updates the encumbrance penalties shown by player::disp_info() in character.cpp so that they show the correct penalties instead of what was presumably true in some earlier version.

### CODE REVIEW

File: src/player.cpp [hunk 1]
```diff
@@ -2329,17 +2329,82 @@ std::string reload_cost_text(int moves)
...
+    if (body_part == bp_leg_r) { leg_r_encumbrance = 0; }
+    float previous_fall_damage = 1.0f;
+    float dex_dodge = player->dex_cur / 2 + player->get_skill_level( skill_dodge );
+    dex_dodge -= ( ( ( leg_l_encumbrance + leg_r_encumbrance ) / 2 ) + ( torso_encumbrance / 1 ) ) / 10;
+    dex_dodge = std::max( 0.0f, dex_dodge );
+    previous_fall_damage *= (100.0f - (dex_dodge * 4.0f)) / 100.0f;
+    if( player->has_trait("PARKOUR") ) {
+        previous_fall_damage *= 2.0f / 3.0f;
+    }
+    return previous_fall_damage;
```
Comment (codemime) [inline src/player.cpp]: This approach doesn't seem to be right. You copy-pasted `player::fall_damage_mod()` and made your tweaks. A more appropriate solution could look somewhat like this:
```c++
float player::fall_damage_mod( int torso_encumb, int legs_encumb ) const {
    ...
}

float player::fall_damage_mod() const {
    return fall_damage_mod( encumb( bp_torso ), encumb( bp_leg_l ) + encumb( bp_leg_r ) );
}
```
Or another (better encapsulated) solution:
```c++
float player::fall_damage_mod_except( const std::bitset<num_bp> &excluded_bps ) const {
    ...
    const int torso_encumb = ( excluded_bps.test( mfb( bp_torso ) ) ) ? encumb( bp_torso ) : 0;
    ...
}

float player::fall_damage_mod() const {
    return fall_damage_mod_except( {} );
}
```
Then when you want to exclude legs from evaluation:
```c++
fall_damage_mod_except( mfb( bp_leg_l ) | mfb( bp_leg_r ) );

```

File: src/player.cpp [hunk 2]
```diff
@@ -2329,17 +2329,82 @@ std::string reload_cost_text(int moves)
...
+}
+
+std::string fall_damage_penalty_text(double unencumbered, double actual)
+{
+    return string_format( _( "Falling damage increases from %.0f%% to %.0f%%. " ), unencumbered, actual );
```
Comment (codemime) [inline src/player.cpp]: Should display a relative integer value( `%+d%%` )

File: src/player.cpp [hunk 3]
```diff
@@ -2977,47 +3042,67 @@ Strength - 4;    Dexterity - 4;    Intelligence - 4;    Perception - 4"));
...
-                s += string_format(_("Perception %+d when checking traps or firing ranged weapons;\n"
-                                     "Perception %+.1f when throwing items."),
-                                   -(encumb(bp_eyes) / 10),
-                                   double(-(encumb(bp_eyes) / 10)) / 2);
-            } else if (line == 3) { //Eyes
-                s += _("Covering your mouth will make it more difficult to breathe and catch your breath.");
+            } else if (line == 2) { //Eyes
+                s += string_format( _( "Perception %+d when checking traps.\n" ), -encumb(bp_eyes) / 10 );
+                s += ranged_penalty_text( -6 * encumb( bp_eyes ) );
+                s += "\n";
```
Comment (codemime) [inline src/player.cpp]: Can be replaced with a char `'\n'` (and better moved into the function). Here and below.

File: src/player.cpp [hunk 4]
```diff
@@ -2977,47 +3042,67 @@ Strength - 4;    Dexterity - 4;    Intelligence - 4;    Perception - 4"));
...
+            if (line == 0) { //Torso
                 const int melee_roll_pen = std::max( -encumb( bp_torso ), -80 );
                 s += string_format( _("Melee attack rolls %+d%%; "), melee_roll_pen );
                 s += dodge_skill_text( - (encumb( bp_torso ) / 10));
                 s += swim_cost_text( (encumb( bp_torso ) / 10) * ( 80 - get_skill_level( skill_swimming ) * 3 ) );
                 s += melee_cost_text( encumb( bp_torso ) );
-            } else if (line == 1) { //Torso
+                s += thrown_cost_text( 2 * encumb( bp_torso ) );
+                s += fall_damage_penalty_text( calc_old_fall_damage_for_encumbrance_penalty( this, bp_torso ) * 100,
+                                               player::fall_damage_mod() * 100 );
```
Comment (codemime) [inline src/player.cpp]: Alot of duplication here. These two lines occur three times and the only thing that varies is the `body_part`. `fall_damage_penalty_text()` should accept it as an argument.

File: src/player.cpp [hunk 5]
```diff
@@ -2329,17 +2329,82 @@ std::string reload_cost_text(int moves)
...
                                     moves ),
                           moves );
 }
 
+std::string thrown_cost_text(int moves)
+{
+    return string_format( ngettext( "Throwing costs %+d movement point. ",
+                                    "Throwing costs %+d movement points. ",
+                                    moves ),
+                          moves );
```
Comment (codemime) [inline src/player.cpp]: This line break is sloppy. Why do you copy it from the existing code?

File: src/player.cpp [hunk 6]
```diff
@@ -2977,47 +3042,67 @@ Strength - 4;    Dexterity - 4;    Intelligence - 4;    Perception - 4"));
...
+                s += throwing_bonus_text( -encumb(bp_eyes) / 10 );
+            } else if (line == 3) { //Mouth
+                s += stamina_regen_cost_text( std::min( 9, -encumb(bp_mouth) / 10 ) );
+                s += string_format( _("Shouting volume %+d."), -encumb( bp_mouth ) * 3 / 2 );
             } else if (line == 4) { //Left Arm
-                s += _("Arm encumbrance affects stamina cost of melee attacks and accuracy with ranged weapons.");
+                s += melee_stamina_cost_text( encumb( bp_arm_l ) / 5.0f);
+                s += "\n";
+                s += ranged_penalty_text( 3 * encumb( bp_arm_l ) );
+                s += "(both per arm)";
```
Comment (codemime) [inline src/player.cpp]: This remark is redundant when body parts are combined.

**Comment by Crispy6** [comment]
Thanks for the feedback.

I didn't want to touch player::fall_damage_mod(), so the code is more readable if the function I added mirrors that. I agree that it would be better for code to be shared between the encumbrance penalty display and calculation, but I have no intention of making that change to only the fall damage. I'd be happy to change all of it if I can get the green light to do so?

I'll try putting all of the newlines in the _cost_text functions and making sure the more crowded body parts still display well.

I'll remove all the text about per arm, per leg, etc., but basically the code to combine pairs of hands, arms, legs, and feet is a complete mess here, because it only drops duplicate body parts instead of changing the combined line in any way. I'll look into how much work it would be to fix that.

For the `%+d%%`, I tested that, and my code worked correctly while that code did not. The two variables being printed are doubles, so displaying them as decimal integers can print nine digits of garbage.

**Comment by Crispy6** [comment]
I'll consult all the style guides I can find for the line breaks and indenting on those long functions, too. I copied the existing one for consistency, since it looked indented correctly to me.

**Comment by Crispy6** [comment]
Putting the newlines into the _cost_text functions doesn't work either, since some body parts were easier to read with newlines and other body parts would be too many lines that way. I'll take all of them out, to bias towards pretty code and away from what I actually wanted displayed onscreen.

**Comment by codemime** [comment]
> I didn't want to touch player::fall_damage_mod(), so the code is more readable if the function I added mirrors that. 

No, it's not more readable, it's more duplicated and hence less readable. Don't be afraid of modifying headers, just leave the declaration of `player::fall_damage_mod()` intact to prevent unnecessary impacts.

> but basically the code to combine pairs of hands, arms, legs, and feet is a complete mess here

Not only there. My eyes bleed when I look at `player::disp_info()`, but it shouldn't stop from gradually changing it.

> For the %+d%%, I tested that, and my code worked correctly while that code did not. The two variables being printed are doubles, so displaying them as decimal integers can print nine digits of garbage.

You should use an integer there, of course. I just didn't state the obvious thing. You calculate an integer percent and write it that way.

**Comment by Crispy6** [comment]
> You should use an integer there, of course. I just didn't state the obvious thing. You calculate an integer percent and write it that way.
player::fall_damage_mod() returns a float, and as I said I was choosing not to modify it at this time. How is a cast to integer better than formatting a float for display?

**Comment by codemime** [comment]
> How is a cast to integer better than formatting a float for display?

That's what I meant:

```c++
std::string fall_damage_penalty_text( const player &p, const std::bitset<num_bp> &excluded_bps  )
{
    const int diff = 100 * ( p.fall_damage_mod() - p.fall_damage_mod_except( excluded_bps ) );
    return string_format( _( "Falling damage increases by %+d%%.  " ), diff );
}

```

**Comment by Crispy6** [comment]
Your encumbrance message above is misleading. Increases by x% of what? I printed both to make it clear how much the fall damage was actually increasing.

I'll close this PR, please nitpick somebody else's code.

**Comment by codemime** [comment]
That wasn't nitpicking and I certainly didn't mean to offend. Sorry if I did.

**Pull request closed without merging**